### ISSUE TITLE
UIStackView as Card content doesn't show it's contents

### ISSUE THREAD

**Issue by ToonvanStrijp** [CosmicMind/Material#974]
I'm trying to set the UIStackView as my card content however the size seems to work, the content of the UIStackView isn't shown.

I'm using the following code:
```
import UIKit
import Material

class LoginViewController: UIViewController{
    fileprivate var card: Card!

    fileprivate var toolbar: Toolbar!
    fileprivate var moreButton: IconButton!


    fileprivate var contentView: UIStackView!

    fileprivate var bottomBar: Bar!
    fileprivate var dateLabel: UILabel!
    fileprivate var favoriteButton: FlatButton!
    fileprivate var defaults: UserDefaults!

    fileprivate var usernameField: TextField!
    fileprivate var passwordField: TextField!

    /// A constant to layout the textFields.
    fileprivate let constant: CGFloat = 32

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        UIApplication.shared.statusBarStyle = .lightContent
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor(red: 177/255, green: 223/255, blue: 247/255, alpha: 1.0)

        //Looks for single or multiple taps.
        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(LoginViewController.dismissKeyboard))
        view.addGestureRecognizer(tap)

        defaults = UserDefaults.standard;

        prepareDateLabel()
        prepareFavoriteButton()
        prepareMoreButton()
        prepareToolbar()
        prepareEmailField()
        preparePasswordField()
        prepareContentView()
        prepareBottomBar()
        prepareCard()
    }

    /// Handle the resign responder button.
    @objc
    internal func login(button: UIButton) {
//        view.endEditing(true)
//        showLoadingIndicator()
//        api.login(email: usernameField.text!, password: passwordField.text!) { (login) in
//            self.hideLoadingIndicator()
//            print(login.message)
//            if(login.success == true){
//                self.defaults.set(login.token, forKey: "token")
//                let toolbarController = AppSnackbarController(rootViewController: MainViewController(rootViewController: ScannerViewController()))
//                let viewController = AppNavigationDrawerController(rootViewController: toolbarController)
//                viewController.modalTransitionStyle = .crossDissolve
//                self.present(viewController, animated: true, completion: nil)
//            }else{
//                self.showSnackbar(message: login.message)
//            }
//        }
    }

    @objc func dismissKeyboard() {
        //Causes the view (or one of its embedded text fields) to resign the first responder status.
        view.endEditing(true)
    }

}

extension LoginViewController{

    fileprivate func prepareDateLabel() {
        dateLabel = UILabel()
        dateLabel.font = RobotoFont.regular(with: 12)
        dateLabel.textColor = Color.blue.base
        dateLabel.text = "Wachtwoord vergeten?"
    }

    fileprivate func prepareFavoriteButton() {
        favoriteButton = FlatButton(title: "Login")
        favoriteButton.pulseAnimation = .pointWithBacking
    }

    fileprivate func prepareMoreButton() {
        moreButton = IconButton(image: Icon.cm.moreVertical, tintColor: Color.grey.base)
    }

    fileprivate func prepareToolbar() {
        toolbar = Toolbar(rightViews: [])

        toolbar.title = "Aanmelden"
        toolbar.titleLabel.textAlignment = .left

        toolbar.detail = "Vul hier je gebruikersnaam en wachtwoord"
        toolbar.detailLabel.textAlignment = .left
        toolbar.detailLabel.textColor = Color.grey.base
    }

    fileprivate func prepareContentView() {
        contentView = UIStackView()
        contentView.axis = UILayoutConstraintAxis.vertical
        contentView.distribution = UIStackViewDistribution.equalSpacing
        contentView.alignment = UIStackViewAlignment.fill
        contentView.spacing = 0
        contentView.translatesAutoresizingMaskIntoConstraints = false
        contentView.addArrangedSubview(usernameField)
        contentView.addArrangedSubview(passwordField)
    }

    fileprivate func prepareEmailField() {
        usernameField = TextField()
        usernameField.autocapitalizationType = .none
        usernameField.textColor = Color.black
        usernameField.dividerNormalColor = Color.blue.base
        usernameField.dividerActiveColor = Color.blue.base
        usernameField.placeholderNormalColor = Color.blue.base
        usernameField.placeholderActiveColor = Color.blue.base
        usernameField.placeholder = "Username"
        usernameField.isClearIconButtonEnabled = true
    }

    fileprivate func preparePasswordField() {
        passwordField = TextField()
        passwordField.textColor = Color.black
        passwordField.dividerNormalColor = Color.blue.base
        passwordField.dividerActiveColor = Color.blue.base
        passwordField.placeholderNormalColor = Color.blue.base
        passwordField.placeholderActiveColor = Color.blue.base
        passwordField.placeholder = "Password"
        passwordField.clearButtonMode = .whileEditing
        passwordField.isVisibilityIconButtonEnabled = true

        passwordField.visibilityIconButton?.tintColor = Color.white.withAlphaComponent(passwordField.isSecureTextEntry ? 0.38 : 0.54)
    }

    fileprivate func prepareBottomBar() {
        bottomBar = Bar()

        bottomBar.leftViews = [dateLabel]
        bottomBar.rightViews = [favoriteButton]
    }

    fileprivate func prepareCard() {
        card = Card()

        card.toolbar = toolbar
        card.toolbarEdgeInsetsPreset = .square3
        card.toolbarEdgeInsets.bottom = 0
        card.toolbarEdgeInsets.right = 8

        card.contentView = contentView
        card.contentViewEdgeInsetsPreset = .wideRectangle3

        card.bottomBar = bottomBar
        card.bottomBarEdgeInsetsPreset = .wideRectangle2

        view.layout(card).width(300).center()
    }
}
```
<img width="598" alt="screen-shot-2017-11-18-at-16 24 50" src="https://user-images.githubusercontent.com/11893063/32989933-07bc5d76-cd20-11e7-93de-41cb47608d0a.png">

**Comment by danieldahan** [comment]
I will be taking a look at this later today. Thank you!

**Comment by joeljfischer** [comment]
I'm seeing this as well, I'd like to have a few vertically stacked controls in my content view, but it doesn't appear possible. I was attempting to get a custom view in there as well, but it wasn't working properly either.

**Comment by danieldahan** [comment]
@joeljfischer Can you send over your code setup, sample project. That will help us. Thank you!

**Comment by danieldahan** [comment]
@ToonvanStrijp @adamdahan Let's replicate this and make Material 3 solves this issue.

**Comment by joeljfischer** [comment]
My code is essentially the same as @ToonvanStrijp.

```swift
class TaskTableViewCell: UITableViewCell {
    @IBOutlet weak var card: Card!
    private var cardToolbar: Toolbar!
    private var cardContentView: UIStackView!
    
    override func awakeFromNib() {
        super.awakeFromNib()

        backgroundColor = Color.white
    }

    func updateUI() {
        prepareCardContentView()
        prepareCardToolbar()
        prepareCard()
    }
}

extension TaskTableViewCell {
    private func prepareCardContentView() {
        let notesLabel = UILabel()
        notesLabel.font = UIFont.systemFont(ofSize: 16.0)
        notesLabel.numberOfLines = 4
        notesLabel.text = "Some text"

        let stackView = UIStackView(arrangedSubviews: [notesLabel])
        stackView.alignment = .fill
        stackView.axis = .vertical
        stackView.distribution = .equalSpacing
        stackView.spacing = 12.0

        cardContentView = stackView
    }

    private func prepareCardToolbar() {
        cardToolbar = Toolbar()
        cardToolbar.backgroundColor = .clear
        cardToolbar.title = "Some title"
        cardToolbar.titleLabel.textAlignment = .natural
    }

    private func prepareCard() {
        card.backgroundColor = Color.grey.lighten4
        card.depthPreset = .depth2
        card.toolbar = cardToolbar
        card.toolbarEdgeInsetsPreset = .square3
        card.toolbarEdgeInsets.bottom = 0
        card.toolbarEdgeInsets.right = 8

        card.contentView = cardContentView
        card.contentViewEdgeInsetsPreset = .wideRectangle3
    }
}
```

**Comment by joeljfischer** [comment]
I appear to no longer have the code for the second issue, using a custom view, but the issue was essentially the same. I just made a custom view, centered the label within it using `layout(label).center()` and passed it like the stack view above.

In the view debugger, the label was there, but the custom view / stack view has a zero height.

**Comment by danieldahan** [comment]
Okay we will investigate this. Make sure that a height is being set for the `contentView` area, otherwise it may not know how to lay itself out.

**Comment by joeljfischer** [comment]
Since it's auto layout, that should be set automatically, right? Like it
would be if it were just a label?

**Comment by danieldahan** [comment]
Not necessarily. It would need to know to layout itself out so that the height may be obtained. To test this, can you try setting the stack view layoutIfNeeded prior to setting it? Like so: 

```swift
stackView.setNeedsLayout()
stackView.layoutIfNeeded()

card.contentView = stackView
```

If that works, I can make an update to Material that will fix your issue. Thank you!

**Comment by joeljfischer** [comment]
That did not work for me 😞 , I had also tried that with the custom view and it didn't work there either.

**Comment by danieldahan** [comment]
Then please send over your project for me to test :) And I am sure we will have a fix.

**Comment by joeljfischer** [comment]
@danieldahan I pushed up a quick project to show the issue: https://github.com/joeljfischer/MaterialBug974

Thanks for the support!

**Comment by danieldahan** [comment]
@joeljfischer excellent, I will take a look. Thank you!

**Comment by danieldahan** [comment]
@joeljfischer After looking at the sample project, there are a few issues: 

1. Setting constraints in IB causes issues when you set use `view.layout`, which sets conflicting constraints. So delete the ones in IB. 

2. The UIStackView has no concept of height, so you need to set the height to some value, or set the layout of the label to generate a height for the stack view. 

3. Ge rid of the stack view and use a container view with a size and then go from there. 

4. Use the label itself, like so: 

```swift
let label = UILabel()
label.font = UIFont.systemFont(ofSize: 16.0)
label.numberOfLines = 4
label.text = "text, text, text, and text"
card.contentView = label
```

If you need any further help, let me know. In the new site we will write a tutorial for this to better help in the future. 

Thank you!

@adamdahan ^

**Comment by joeljfischer** [comment]
Thanks for the support again. When you do that tutorial, maybe some of my further findings can help.

1. This was just a mis-step in the example, removing them does not affect the issue at hand.
2. I'll come back to this
3. I've tested this and it has the same issues as before.
4. Sorry, I think I miscommunicated the issue. I wish to use multiple views (e.g. multiple labels, or a label and another view) and let the label height determine the height of the content view of the card (e.g. if it has 3 lines, the content view is larger than if it only has one line).

Coming back to (2), this seems to be the heart of the issue. A label has an intrinsic content size, and the stack view, even with arranged subviews present, does not. So it needs to be manually given a height.

Just using a height anchor doesn't seem to do the job `stackView.heightAnchor.constraint(equalToConstant: 128.0)` doesn't work.

This, however, did work: `stackView.bounds = CGRect(x: 0, y: 0, width: card.bounds.width, height: label.intrinsicContentSize.height)` or more completely:

```swift
let stackViewIntrinsicHeight = stackView.arrangedSubviews.reduce(0, { result, view in
    result + view.intrinsicContentSize.height
})
stackView.bounds = CGRect(x: 0, y: 0, width: card.bounds.width, height: stackViewIntrinsicHeight)
```

This would, however, only work with views that _have_ an intrinsic content size, so perhaps an _even more complete_ method would check for height anchors or bounds sizes? In any case, since my use case is to be have a label and a custom view below that, so I'll have to know the height of my custom view somehow as well – though I haven't gotten to that part yet.

Hope that helps in writing the tutorial / docs, and thanks for the help.
